### ISSUE TITLE
Dual use modifier/normal keys

### ISSUE THREAD

**Issue by algernon** [algernon/Akela#56]
Keys that act as a modifier when held, a normal key if tapped. Should support all 8 modifiers, but only one at a time. Needs a wide range because we do not touch the `rawKey part`...

Encoding this in the keycode will be the toughest part.

**Comment by algernon** [comment]
Okay, so we have 8 modifiers, so we'll need three bits off the flags, or a range of 8*256, if we want to stick to flags. That's not going to work, we need a smarter way. Using the rawKey area only is not going to work, either, because that limits the amount of normal keys.

**Comment by algernon** [comment]
QMK uses the top three bits. I guess there's no way around that, and this will be one of the big ranges.

**Comment by algernon** [comment]
So, hm. Lets see.

## Scenario 1:

- Press `Ctrl/Esc`: notice that this is a dual-use key, start a timer, set the state bit.
- Immediately release `Ctrl/Esc`: state bit set, timer not timed out yet, register `Esc`, clear state bit, and the timer.

## Scenario 2:

- Press `Ctrl/Esc`: dual key, set the state bit, start timer.
- Keep holding it.
- Notice the timeout, register `Ctrl`, and clear the state bit.
- Release `Ctrl/Esc`: no state bit, timer elapsed, no `Esc` to register.

## Scenario 3:

- Press `Ctrl/Esc`: dual-use key, start a timer, set the state bit.
- Press `z`: register `Ctrl` immediately, clear the state bit.
- Keep holding `Ctrl/Esc`: timer still increases, but on timeout, there is no state bit, so no `Esc` registered. The state bit is only set when the key toggles on, never on hold.

## Summary

- Need a state bit for each modifier.
- Need a separate timer for each modifier. An 8-bit one is enough, possibly a 4-bit one would do, too, but lets go with 8 bits for now. The separate timers are needed so we can have more than one dual-use key active.
- `Ctrl` registers after a timeout, not immediately, and keeps re-registering while pressed, and the state bit is unset. (this behaviour is deferred to other handlers.)
- `Esc` registers only once, on release, when uninterrupted, and within the timeout period.

To build SpaceCadet on this behaviour, the parens repeating would be lost, but that's generally not an issue. However, we'll need to register a shift too, and there's not enough bits for that. So SpaceCadet will need a bit more than `SFT_T(Key_LEFT_PARENS)`, but can be a subclass of `DualUseMod`, or something.

Or perhaps, there could be a knob: whether to keep registering the modifier when the bit is unset, until release. In this case, it would register immediately, too, and we wouldn't even use a timer, as there's no point. But then its an entirely different thing, so... maybe `SpaceCadet` should be completely independent after all.

**Comment by algernon** [comment]
Regarding SpaceCadet: it still needs a timer, so there's a point for the knob. I suppose the best would be to implement the generic dual-use keys, and see if SpaceCadet can be rebuilt on top of that.
